### PR TITLE
gsdx-ogl: Improve accurate blend sprite draw speed

### PR DESCRIPTION
**Pull request by turtleli**
Needs testing - AppVeyor builds for Windows users available when checks complete.

Changes:
 * List of number of non-overlapping consecutive sprite draw calls is generated
 * Non-overlapping consecutive sprite draw calls are now drawn together
 * Plus debug messages to make life easier

Improves performance for accurate blend >= 2 (Medium and above) - See #1093

Note: An initial attempt to use SSE to replace the ifs seemed to slow things down - I lost a few fps on Windows (AVX2 build). Code used (maybe someone has better ideas):
```
GSVector4i all(v[0].XYZ.X, v[0].XYZ.Y, v[1].XYZ.X, v[1].XYZ.Y);
all = all.xyxy().blend(all.zwzw(), all > all.zwxy());
```

### CODE REVIEW

**Comment by gregory38** [comment]
That strange that sse is slower. I mean data is already in vector and branches aren't reputed to be faster. Did you try to just compute min/max without any comparison ? 

@gabest11 if you have some ideas feel free to share :) Note maybe it would be doable to do it in VertexTrace? As a reminder the idea is to generate a list of the number of consecutive primitives that don't overlap (at least for sprite, but could be nice for triangle).

**Comment by gregory38** [comment]
Here a gs dump on virtual fighter. It is interesting because it uses 8192 sprites to draw the screen. It is a good way to profile the overhead of the overlap function. Strangely it seems that AB medium is faster than basic...

https://drive.google.com/file/d/0BzftmrM8nnSoNlRUdHlKRF9iWkk/view?usp=sharing

**Comment by gabest11** [comment]
GSVector4i all(v[0].XYZ.X, v[0].XYZ.Y, v[1].XYZ.X, v[1].XYZ.Y);

This one could be done better, don't refer to members of a vector, use shuffles and unpacks to make "all". I always dump the generated assembly when working with vectors and check for register spills and unnecessary memory refs to the members. It only makes sense to do it this way, if the compiler creates pure vector instructions.

Then runion has a test for empty or reversed rectangles, which is not needed here, xy and zw are already sorted. Just use the i = 0xffff case from it.

About array addressing. It is always the most easy on the compiler to use one pointer and change the index throughout the whole function. Have a base pointer to the vertices defined at the beginning and use i, j with absolute values (j = i + 2). Then you can even get rid of remaining_count and just use count there, too.

For many sprites this loop could add a lot of pre-processing for each drawing, maybe check for an upper limit?

Vertex trace only works on individual vertices, not prims.

**Comment by gregory38** [comment]
```
Then runion has a test for empty or reversed rectangles, which is not needed here, xy and zw are already sorted. Just use the i = 0xffff case from it.
```
Good point. It would help.

```
For many sprites this loop could add a lot of pre-processing for each drawing, maybe check for an upper limit?
```
Hum. Normally it is doable. The above dump uses various 8K sprites draw call. It is still faster to check the overlapping of a big draw than to split the draw. However, it is likely a good idea to call it when we really need it. The logical isn't optimal here.

```
Vertex trace only works on individual vertices, not prims.
```
Ah yes you're right. I'm silly


@gabest11 
By the way, off-topic, but we ought to discuss one day, Read/Write render target on the (SW?) renderer. Current behavior is wrong because it doesn't properly emulate the texture page buffer. As you know a read/write draw on the same page will act as if the RT was copied at T=0 into a RO texture. However if the texture is sampled in multiple pages, it will trigger a refetch of the texture page, basically we need to have a RW target in this case. Currently you copy the RT so the former case is correctly handled but the latter isn't. It is used on GTA for radiosity effect and Jak&Dexter game for the shadow (albeit could maybe be hacked).

**Comment by gabest11** [comment]
I don't think that single page cache will ever be emulated. Every texture sampling may trigger an interrupt (like a page fault) which replaces the page in the cache, cannot be done fast enough deep inside the core of drawing function. This is probably very expensive even on ps2, that's why they do screen copy in page wide sprites only.

**Comment by gregory38** [comment]
Hum, I can do something on the hardware renderer. At least on some special conditions.

I think we need to first detect the bad draw call for example TBP == FBP and texture is sampled on more than a page. Then would it be possible to uses a RW target with a "flush" between primitives for those draw calls. Otherwise even a slow option will be nice for debug.

**Comment by turtleli** [comment]
Updated (vectorised, runion stuff, array addressing) - Thanks to @gabest11 for the advice.

Vector stuff is now
```
GSVector4i all = GSVector4i(v[i].m[1]).upl16(GSVector4i(v[i+1].m[1])).upl16().xzyw();
all = all.xyxy().blend(all.zwzw(), all > all.zwxy());
```

Some brief testing by replaying the dumps (AVX2 build, Win 10, accurate blend at medium):

| dump | master | without optimise commit | with optimise commit |
| --- | --- | --- | --- |
| Virtua Fighter (non-overlap) | ~219fps | ~212fps | ~217fps |
| Zone of Enders (overlap) | ~39.5fps | ~121fps | ~122fps |

**Comment by gregory38** [comment]
Ah nice. I would suggest to make a runion_orderer function to gsvector to avoid the spreading of the _M_SSE check.

**Comment by turtleli** [comment]
Ok, runion_ordered suggestion done.

File: plugins/GSdx/GSRendererOGL.cpp [hunk 1]
```diff
@@ -606,6 +597,32 @@ void GSRendererOGL::SendDraw(bool require_barrier)
 		ASSERT(GLLoader::found_GL_ARB_texture_barrier);
 		glTextureBarrier();
 		dev->DrawIndexedPrimitive();
+	} else if (m_vt.m_primclass == GS_SPRITE_CLASS) {
+		size_t nb_vertex = (GLLoader::found_geometry_shader) ? 2 : 6;
+
+		GL_PUSH("Split the draw");
+
+#if defined(ENABLE_OGL_DEBUG)
```
Comment (gregory38) [inline plugins/GSdx/GSRendererOGL.cpp]: Use _DEBUG here instead. OGL_DEBUG is enabled on dev build and it is already slow enough.

File: plugins/GSdx/GSRendererOGL.cpp [hunk 2]
```diff
@@ -606,6 +597,32 @@ void GSRendererOGL::SendDraw(bool require_barrier)
 		ASSERT(GLLoader::found_GL_ARB_texture_barrier);
 		glTextureBarrier();
 		dev->DrawIndexedPrimitive();
+	} else if (m_vt.m_primclass == GS_SPRITE_CLASS) {
+		size_t nb_vertex = (GLLoader::found_geometry_shader) ? 2 : 6;
+
+		GL_PUSH("Split the draw");
```
Comment (gregory38) [inline plugins/GSdx/GSRendererOGL.cpp]: Could you add something to seperate it from the other split draw code. For example  a `(SPRITE)` or `bounding box algo`.

File: plugins/GSdx/GSRendererOGL.cpp [hunk 3]
```diff
@@ -606,6 +597,32 @@ void GSRendererOGL::SendDraw(bool require_barrier)
...
+
+#if defined(ENABLE_OGL_DEBUG)
+		// Some code to check how draw code is split.
+		map<size_t, size_t> frequency;
+		for (auto it = m_drawlist.begin(); it != m_drawlist.end(); ++it)
```
Comment (gregory38) [inline plugins/GSdx/GSRendererOGL.cpp]: Now that we droped VS2012, we could also use the new range C++11 syntax. It is up to you. Same for the loop below. By the way, maybe be better to use a reference `auto& it`

**Comment by gregory38** [comment]
Except the minor nitpick above. Code is good and reviewed by me :+1:

**Comment by turtleli** [comment]
Suggested changes done.

**Comment by gregory38** [comment]
Ok. Let's merge it then.

**Pull request merged**